pkg/neotest/coverage.go

```diff
@@ -0,0 +1,157 @@
+package neotest
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/nspcc-dev/neo-go/pkg/compiler"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm"
+	"github.com/nspcc-dev/neo-go/pkg/vm/opcode"
+)
+
+var rawCoverage = make(map[util.Uint160]*scriptRawCoverage)
+
+var enabled bool
+var coverProfile = ""
+
+type scriptRawCoverage struct {
+	debugInfo      *compiler.DebugInfo
+	offsetsVisited []int
+}
+
+type coverBlock struct {
+	startLine uint // Line number for block start.
+	startCol  uint // Column number for block start.
+	endLine   uint // Line number for block end.
+	endCol    uint // Column number for block end.
+	stmts     uint // Number of statements included in this block.
+	counts    uint
+}
+
+type documentName = string
+
+func isCoverageEnabled() bool {
+	if enabled {
+		return true
+	}
+	const coverProfileFlag = "test.coverprofile"
+	flag.VisitAll(func(f *flag.Flag) {
+		if f.Name == coverProfileFlag && f.Value != nil {
+			enabled = true
+			coverProfile = f.Value.String()
+		}
+	})
+	if enabled {
+		// this is needed so go cover tool doesn't overwrite
+		// the file with our coverage when all tests are done
+		err := flag.Set(coverProfileFlag, "")
+		if err != nil {
+			panic(err)
+		}
+	}
+	return enabled
+}
+
+func coverageHook() vm.OnExecHook {
+	return func(scriptHash util.Uint160, offset int, opcode opcode.Opcode) {
+		if cov, ok := rawCoverage[scriptHash]; ok {
+			cov.offsetsVisited = append(cov.offsetsVisited, offset)
+		}
+	}
+}
```

Comment:
Could you explain why `coverageHook` is a function? Can't we have a global variable that contains an instance of `vm.OnExecHook`? It's read-only anyway, no one modifies it.